### ISSUE TITLE
wasm2wat needed_dynlibs error

### ISSUE THREAD

**Issue by syrusakbary** [WebAssembly/wabt#1429]
While trying to fix one issue in Wasmer, we realized on a bug that `wabt` have, but we have no idea what can be causing it.

```bash
$ wasm2wat lib/c-api/tests/assets/hello_wasm.wasm
0000018: error: unable to read u32 leb128: needed_dynlibs
```

```bash
$ wasm-validate lib/c-api/tests/assets/hello_wasm.wasm
0000018: error: unable to read u32 leb128: needed_dynlibs
```

The [`hello_wasm.wasm` file](https://github.com/wasmerio/wasmer/blob/master/lib/runtime-c-api/tests/assets/hello_wasm.wasm) seem to be valid (according to other validators), so we are not really sure what's going on.

The wasm file was generated using rustc with `wasm32-unknown-unknown` target for this source:
https://github.com/wasmerio/wasmer/blob/master/lib/runtime-c-api/tests/assets/return_hello.rs

**Comment by binji** [comment]
Looks like it is trying to read the `dylink` custom section, and erroring out because it's missing that field. @sbc100 looks like you added this, do you know what may be happening?

**Comment by sbc100** [comment]
Thanks, I'll take a look.

**Comment by sbc100** [comment]
Hmm, what version of llvm was used to compile that file?  It looks like that I added the extrea leb to that section back in https://reviews.llvm.org/D55613, which I think means it was in llvm 8 and above.

My guess is that other validators ignore this section completely, and I guess maybe(?) wasm-validate should too?

**Comment by sbc100** [comment]
Also, I don't think `wasm32-unknown-unknown` should be producing dynamic libraries should it?  Does it even support the `-shared` flag?

**Comment by syrusakbary** [comment]
> Hmm, what version of llvm was used to compile that file? 
Unfortunately we don't have the script that we used to generate the file, so we can't know for sure. It could be that was the rustc version we were using was using llvm-7?

**Comment by sbc100** [comment]
Yes I think wasm-ld from llvm 7 would produce such a dylink section, but even then only if linked as a shared library with `-shared` or `-pie` I believe.

Either way we should re-visit how we validate this custom section.
